File: src/utils/cache.ts
/**
 * Simple in-memory cache with TTL for better scalability
 * Reduces redundant API calls and improves performance
 */

interface CacheEntry<T> {
  data: T;
  timestamp: number;
  ttl: number; // Time to live in milliseconds
}

export class Cache {
  private static instance: Cache;
  private cache: Map<string, CacheEntry<any>>;
  private maxSize: number;

  private constructor(maxSize: number = 1000) {
    this.cache = new Map();
    this.maxSize = maxSize;
    
    // Clean up expired entries every minute
    setInterval(() => this.cleanup(), 60000);
  }

  static getInstance(maxSize?: number): Cache {
    if (!Cache.instance) {
      Cache.instance = new Cache(maxSize);
    }
    return Cache.instance;
  }

  /**
   * Get value from cache
   */
  get<T>(key: string): T | null {
    const entry = this.cache.get(key);
    
    if (!entry) {
      return null;
    }

    // Check if expired
    const now = Date.now();
    if (now - entry.timestamp > entry.ttl) {
      this.cache.delete(key);
      return null;
    }

    return entry.data as T;
  }

  /**
   * Set value in cache
   */
  set<T>(key: string, data: T, ttl: number = 60000): void {
    // Evict oldest entries if cache is full
    if (this.cache.size >= this.maxSize) {
      this.evictOldest();
    }

    this.cache.set(key, {
      data,
      timestamp: Date.now(),
      ttl,
    });
  }

  /**
   * Delete value from cache
   */
  delete(key: string): void {
    this.cache.delete(key);
  }

  /**
   * Clear all cache
   */
  clear(): void {
    this.cache.clear();
  }

  /**
   * Get cache size
   */
  size(): number {
    return this.cache.size;
  }

  /**
   * Clean up expired entries
   */
  private cleanup(): void {
    const now = Date.now();
    for (const [key, entry] of this.cache.entries()) {
      if (now - entry.timestamp > entry.ttl) {
        this.cache.delete(key);
      }
    }
  }

  /**
   * Evict oldest entry
   */
  private evictOldest(): void {
    let oldestKey: string | null = null;
    let oldestTimestamp = Infinity;

    for (const [key, entry] of this.cache.entries()) {
      if (entry.timestamp < oldestTimestamp) {
        oldestTimestamp = entry.timestamp;
        oldestKey = key;
      }
    }

    if (oldestKey) {
      this.cache.delete(oldestKey);
    }
  }
}

/**
 * Cache decorator for async functions
 */
export function cached<T extends (...args: any[]) => Promise<any>>(
  fn: T,
  options: {
    keyGenerator?: (...args: Parameters<T>) => string;
    ttl?: number;
    cache?: Cache;
  } = {}
): T {
  const {
    keyGenerator = (...args) => JSON.stringify(args),
    ttl = 60000, // 1 minute default
    cache = Cache.getInstance(),
  } = options;

  return (async (...args: Parameters<T>): Promise<ReturnType<T>> => {
    const key = `${fn.name}_${keyGenerator(...args)}`;
    
    // Try to get from cache
    const cached = cache.get<ReturnType<T>>(key);
    if (cached !== null) {
      return cached;
    }

    // Execute function and cache result
    const result = await fn(...args);
    cache.set(key, result, ttl);
    
    return result;
  }) as T;
}

File: flask_backend/utils/cache.py
"""
Simple in-memory cache with TTL for better scalability
Reduces redundant database queries and improves performance
"""
from datetime import datetime, timedelta
from typing import Any, Optional, Callable
import threading
import time

class CacheEntry:
    def __init__(self, data: Any, ttl: int = 60):
        self.data = data
        self.timestamp = datetime.now()
        self.ttl = timedelta(seconds=ttl)

    def is_expired(self) -> bool:
        return datetime.now() - self.timestamp > self.ttl


class Cache:
    """Thread-safe in-memory cache with TTL"""
    
    def __init__(self, max_size: int = 1000):
        self.cache: dict[str, CacheEntry] = {}
        self.max_size = max_size
        self.lock = threading.Lock()
        self._cleanup_interval = 300  # 5 minutes
        self._last_cleanup = time.time()

    def get(self, key: str) -> Optional[Any]:
        """Get value from cache"""
        with self.lock:
            self._cleanup_if_needed()
            
            entry = self.cache.get(key)
            if entry is None:
                return None
            
            if entry.is_expired():
                del self.cache[key]
                return None
            
            return entry.data

    def set(self, key: str, data: Any, ttl: int = 60) -> None:
        """Set value in cache"""
        with self.lock:
            self._cleanup_if_needed()
            
            # Evict oldest if cache is full
            if len(self.cache) >= self.max_size and key not in self.cache:
                self._evict_oldest()
            
            self.cache[key] = CacheEntry(data, ttl)

    def delete(self, key: str) -> None:
        """Delete value from cache"""
        with self.lock:
            self.cache.pop(key, None)

    def clear(self) -> None:
        """Clear all cache"""
        with self.lock:
            self.cache.clear()

    def _cleanup_if_needed(self) -> None:
        """Clean up expired entries if enough time has passed"""
        now = time.time()
        if now - self._last_cleanup < self._cleanup_interval:
            return
        
        self._last_cleanup = now
        expired_keys = [
            key for key, entry in self.cache.items()
            if entry.is_expired()
        ]
        for key in expired_keys:
            del self.cache[key]

    def _evict_oldest(self) -> None:
        """Evict oldest entry"""
        if not self.cache:
            return
        
        oldest_key = min(
            self.cache.keys(),
            key=lambda k: self.cache[k].timestamp
        )
        del self.cache[oldest_key]

    def size(self) -> int:
        """Get cache size"""
        with self.lock:
            return len(self.cache)


# Global cache instance
_cache_instance: Optional[Cache] = None


def get_cache() -> Cache:
    """Get global cache instance"""
    global _cache_instance
    if _cache_instance is None:
        _cache_instance = Cache()
    return _cache_instance


def cached(ttl: int = 60):
    """Decorator to cache function results"""
    def decorator(func: Callable) -> Callable:
        cache = get_cache()
        
        def wrapper(*args, **kwargs):
            # Generate cache key from function name and arguments
            key = f"{func.__name__}:{str(args)}:{str(sorted(kwargs.items()))}"
            
            # Try to get from cache
            cached_result = cache.get(key)
            if cached_result is not None:
                return cached_result
            
            # Execute function and cache result
            result = func(*args, **kwargs)
            cache.set(key, result, ttl)
            
            return result
        
        return wrapper
    return decorator

